pkg/scheduler/flavorassigner/podset_reducer.go

```diff
@@ -0,0 +1,92 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package flavorassigner
+
+import (
+	"sort"
+
+	"k8s.io/utils/pointer"
+
+	kueue "sigs.k8s.io/kueue/apis/kueue/v1beta1"
+)
+
+// PodSetReducer helper structure used to gradually walk down
+// from PodSets[*].Count to *PodSets[*].MinimumCount.
+type PodSetReducer[R any] struct {
+	podSets    []kueue.PodSet
+	fullCounts []int32
+	deltas     []int32
+	totalDelta int32
+	fits       func([]int32) (R, bool)
+}
+
+func NewPodSetReducer[R any](podSets []kueue.PodSet, fits func([]int32) (R, bool)) *PodSetReducer[R] {
+	psr := &PodSetReducer[R]{
+		podSets:    podSets,
+		deltas:     make([]int32, len(podSets)),
+		fullCounts: make([]int32, len(podSets)),
+		fits:       fits,
+	}
+
+	for i := range psr.podSets {
+		ps := &psr.podSets[i]
+		psr.fullCounts[i] = ps.Count
+
+		d := ps.Count - pointer.Int32Deref(ps.MinCount, ps.Count)
+		psr.deltas[i] = d
+		psr.totalDelta += d
+	}
+	return psr
+}
+
+func scale(in []int32, out []int32, upFactor int32, downFactor int32) {
+	// this will panic if len(out) < len(in)
+	for i, v := range in {
+		out[i] = v * upFactor / downFactor
+	}
+}
+
+func sub(a, b, out []int32) {
+	for i := range a {
+		out[i] = a[i] - b[i]
+	}
+}
+
+// Find the first biggest set of counts that pass fits(), it's using binary Search
+// so the last call to fits() might not be a successful one
+// Returns nil if no solution was found
+func (psr *PodSetReducer[R]) Search() (R, bool) {
+	var lastGoodIdx int
+	var lastR R
+
+	if psr.totalDelta == 0 {
+		return lastR, false
+	}
+
+	current := make([]int32, len(psr.podSets))
+	idx := sort.Search(int(psr.totalDelta)+1, func(i int) bool {
```

Comment:
For big numbers, I think this approach will skip some values.
That could be fine or it could be surprising to some users.

Given this uncertainty and #820, I think we should really limit the feature to one podset.

Reply:
I spoke too soon: taking the max is enough for 2 pod sets, like seen above.

However, for (1, 2, 3), the sum gives you one more set option than max.

- max
<details>

| search | ps1 | ps2 | ps3 |
| - | - | - | - |
| 0 | 0 | 0 | 0 |
| 1 | 0 | 0 | 1 |
| 2 | 0 | 1 | 2 |
| 3 | 1 | 2 | 3 |

</details>

- sum
<details>

| search | ps1 | ps2 | ps3 |
| - | - | - | - |
| 0 | 0 | 0 | 0 |
| 1 | 0 | 0 | 0 |
| 2 | 0 | 0 | 1 |
| 3 | 0 | 1 | 1 |
| 4 | 0 | 1 | 2 |
| 5 | 0 | 1 | 2 |
| 6 | 1 | 2 | 3 |

</details>

Reply:
Still the question remains about whether this set of tries is acceptable to users. I prefer we do not commit to an algorithm if we don't have user feedback. Then, we can start for support for single podset only.

Reply:
I have tried it early on bu depending on the numbers,  if we go with max we can end up lousing some combinations, starting from 3 and 5 
```
max
0 + 0 = 0
0 + 1 = 1
1 + 2 = 3
1 + 3 = 4
2 + 4 = 6

sum
0 + 0 = 0
0 + 1 = 1
1 + 1 = 2
1 + 2 = 3
1 + 3 = 4
2 + 3 = 5
2 + 4 = 6
```
so we will not test (1,1) and (2,3)

since the binary search is O(log n) we could ignore the number of duplicates.

https://go.dev/play/p/6Hy734t-Cgx

Reply:
Yeah, I'm fine with duplicates, but is the set of combinations good enough?
I rather not go through that rabbit hole before we have more clarity about podset and rayjob.
We could check how Ray does autoscaling or preemption. But then we should still defer that for later and stick to one podset for this PR.

Reply:
It seems that most people prefer we don't commit to multiple podsets yet. Can you strip it down?

Reply:
the webhook part is added

Reply:
we'll go deeper in the ray implementation with #783